### PR TITLE
Eliminates expensive librecords and malloc on drain

### PR DESCRIPTION
**Pull request by zwoop**

### CODE REVIEW

**Comment by zwoop** [comment]
Note that these expensive (locks) librecords lookups happen on every request. So, both locking and the malloc() (new()) is eliminated this way.

**Comment by maskit** [comment]
[approve ci autest]

File: proxy/ProxyClientSession.h [hunk 1]
```diff
@@ -69,6 +69,9 @@ struct ProxyError {
   uint32_t code       = 0;
 };
 
+// A little ugly, but this global is tracked by traffic_server.
+extern bool ts_is_draining;
```
Comment (bryancall) [inline proxy/ProxyClientSession.h]: We should create a naming standard for globals (TS_ or something).

Comment (zwoop) [reply proxy/ProxyClientSession.h]: Agreed, but I couldn't find anything existing :).

**Comment by zwoop** [comment]
@ezelkow1 This also got strange, so if you can take this one, and take a look at the merge conflicts, that'd be great.

**Comment by zwoop** [comment]
Cherry-picked to 8.0.x

**Comment by ezelkow1** [comment]
> @ezelkow1 This also got strange, so if you can take this one, and take a look at the merge conflicts, that'd be great.

On the 7.x backport, this relies too much on new functionality, new file/dir structures, etc. So not really possible to backport to 7.1.7

**Pull request merged**